### PR TITLE
chore(tests): improve EIP-8024 coverage, checklist, and ref-spec pin

### PR DESCRIPTION
**Pull request by spencer-tb**
### Description

Aims to close out EIP-8024 test coverage for the `tests@v21.0.0` tracker by adding fork transition, mainnet, and execution context tests (call frames, set code, initcode), a DUPN stack overflow boundary and empty stack case, a SWAPN full stack sweep, and the jump-into-immediate composite from #1877.

Re pins the reference spec (old pin was a dangling pre merge commit) and replaces the spec.py decode helpers, previously imported from the EELS implementation under test, with the EIP's own reference decoders so decode expectations are implementation independent. Completes the testing checklist.

### Related Issues or PRs

#1877, #2726.

### Checklist

- [x] Ran fast static checks to avoid CI fails, see [Code Standards](https://steel.ethereum.foundation/docs/execution-specs/getting_started/code_standards/) & [Verifying Changes](https://steel.ethereum.foundation/docs/execution-specs/getting_started/verifying_changes/): `just static`
- [x] PR title has the form `<type>(<area>): <title>`, where `<type>` and `<area>` come from an appropriate [`C-<type>`](https://github.com/ethereum/execution-specs/labels?q=C-), respectively [`A-<area>`](https://github.com/ethereum/execution-specs/labels?q=A-), label. The title should match the target squash commit message.

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/amsterdam/eip8024_dupn_swapn_exchange/spec.py [hunk 1]
```diff
@@ -40,12 +35,32 @@ class Spec:
...
+    """
+    Decode the DUPN/SWAPN immediate byte per the EIP-8024 reference code.
 
+    Return n with 17 <= n <= 235.
 
-def decode_single(x: int) -> int:
-    """Decode single with proper typing for tests."""
-    return int(_decode_single(U8(x)))
+    Implemented verbatim from the EIP so test expectations stay
+    independent of the EELS implementation under test.
```
Comment (LouisTsai-Csie) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/spec.py]:
```suggestion
```

Do we need the this? I think adding explanation in the PR description is enough. Same applies to `decode_pair`.

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py [hunk 2]
```diff
@@ -398,3 +408,110 @@ def test_dupn_with_dup1_sequence(
...
 
     state_test(pre=pre, post=post, tx=tx)
+
+
+@EIPChecklist.Opcode.Test.StackOverflow()
+@pytest.mark.parametrize(
+    "stack_height,call_succeeds",
+    [
+        pytest.param(1023, True, id="dupn_fills_stack_to_limit"),
+        pytest.param(1024, False, id="dupn_stack_overflow"),
```
Comment (LouisTsai-Csie) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py]: Nit: Instead of hardcoding the max stack height, use `fork.max_stack_height()`, then updating the parametrization:

```python
@pytest.mark.parametrize(
    "stack_delta,call_succeeds",
    [
        pytest.param(-1, True),
        pytest.param(0, False),
    ],
)
...

stack_height = fork.max_stack_height() + stack_delta
```

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py [hunk 3]
```diff
@@ -398,3 +408,110 @@ def test_dupn_with_dup1_sequence(
...
+            storage={0: 1 if call_succeeds else 0},
+        ),
+    }
+
+    state_test(pre=pre, post=post, tx=tx)
+
+
+@EIPChecklist.Opcode.Test.StackComplexOperations.StackHeights.Zero()
+@EIPChecklist.Opcode.Test.StackUnderflow()
+def test_dupn_empty_stack(
```
Comment (LouisTsai-Csie) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py:459]: Could we test `SWAPN` and `EXCHANGE` on empty stack too?

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_vectors.py [hunk 4]
```diff
@@ -786,6 +786,49 @@ def test_vector_exchange_invalid_0x52(
...
+    PUSH1 0x5b]), parametrized over all three opcodes. JUMPDEST analysis
+    is unchanged by EIP-8024: the byte after the opcode is analyzed as
+    PUSH1, whose data portion masks the 0x5b, so the jump to it must
+    fail. A client that instead masks the opcode's immediate would leave
+    the 0x5b at an instruction boundary, accept the jump, and succeed —
+    diverging on bytecode that is valid before and after the fork.
+    """
+    sender = pre.fund_eoa()
+
+    # Layout: [0-1] PUSH1 5, [2] JUMP, [3] opcode, [4] 0x60, [5] 0x5b
```
Comment (LouisTsai-Csie) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_vectors.py]: Suggested comment layout:
```
00 PUSH1 0x05
02 JUMP
03 <opcode> 0x60
05 JUMPDEST
```

or 

```
00 PUSH1 0x05
02 JUMP
03 <opcode>
04 PUSH1 0x5b
```

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_vectors.py [hunk 5]
```diff
@@ -786,6 +786,49 @@ def test_vector_exchange_invalid_0x52(
...
+    code = Op.PUSH1(5) + Op.JUMP + opcode[b"\x60"] + Op.JUMPDEST
+    expected_bytes = (
+        bytes.fromhex("600556") + bytes(opcode) + bytes.fromhex("605b")
+    )
+    assert bytes(code) == expected_bytes
```
Comment (LouisTsai-Csie) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_vectors.py:819]:
```suggestion
```

I think this is unnecessary?

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py [hunk 6]
```diff
@@ -398,3 +408,110 @@ def test_dupn_with_dup1_sequence(
...
+    forbidden 0x5b immediate halts too), never succeed.
+    """
+    sender = pre.fund_eoa()
+
+    code = Bytecode()
+    code += Op.PUSH1(4)  # Push jump target (position 4, the 0x5b)
+    code += Op.JUMP
+    code += Op.DUPN[b"\x5b"]  # Positions 3-4: DUPN + 0x5b (JUMPDEST)
+    code += Op.PUSH1(3)  # Push jump target (position 3, the DUPN)
+    code += Op.JUMP
```
Comment (LouisTsai-Csie) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py]: The behavior for correct & wrong client implementation are all the same here:

Behavior Comparison between correct vs wrong Implementations -> all abort

- Correct Implementation: Jump to index `0x04` → then jump to index `0x03`, abort since index `0x03` is DUPN with invalid immediate.
- Wrong Implementation 1 (Masks `JUMPDEST` analysis as immediate bytes): Aborts on the first jump, since it incorrectly recognizes `0x5b` as immediate bytes of the first `JUMP`, not as a `JUMPDEST`.
- Wrong Implementation 2 (Executes `DUPN`): If the first `JUMP` succeeds and jumps to `DUPN`, and `DUPN` is recognized as a valid operation, execution continues to the end and jumps back to `DUPN` again. This repeats until either OOG or stack size limit is exceeded.


What do you think about the following?
```
PUSH0 * 17; PUSH1(target) JUMP; [target] DUPN[0x80]; SSTORE(0, 0x42)
```

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_exchange.py [hunk 7]
```diff
@@ -451,6 +458,59 @@ def test_exchange_with_push_sequence(
     state_test(pre=pre, post=post, tx=tx)
 
 
+def test_exchange_full_stack(
```
Comment (LouisTsai-Csie) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_exchange.py:494]: Suggestion to make the test more flexible:

```python
@pytest.mark.parametrize(
    "n,m",
    [
        pytest.param(1, 2, id="adjacent"),
        pytest.param(14, 16, id="max_n"),
        pytest.param(1, 29, id="deepest"),
    ],
)
def test_exchange_full_stack(
    n: int,
    m: int,
    pre: Alloc,
    fork: Fork,
    state_test: StateTestFiller,
) -> None:
    window = Spec.EXCHANGE_MAX_M + 1

    code = Op.PUSH0 * (fork.max_stack_height() - window)
    for depth in range(window, 0, -1):
        code += Op.PUSH1(depth)

    code += Op.EXCHANGE[n, m]
    for slot in range(window):
        code += Op.PUSH1(slot) + Op.SSTORE
    code += Op.STOP

    expected = {slot: slot + 1 for slot in range(window)}
    expected[n], expected[m] = expected[m], expected[n]

    contract_address = pre.deploy_contract(code=code)
    tx = Transaction(to=contract_address, sender=pre.fund_eoa())

    post = {contract_address: Account(storage=expected)}

    state_test(pre=pre, post=post, tx=tx)
```

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_swapn.py [hunk 8]
```diff
@@ -475,3 +485,55 @@ def test_swapn_with_dup1_and_push(
...
+
+
+@EIPChecklist.Opcode.Test.StackComplexOperations.DataPortionVariables.Top()
+@EIPChecklist.Opcode.Test.StackComplexOperations.DataPortionVariables.Middle()
+@pytest.mark.parametrize(
+    "stack_index",
+    [17, 126, 235],
+    ids=lambda x: f"swapn_full_stack_{x}",
+)
+def test_swapn_full_stack(
```
Comment (LouisTsai-Csie) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_swapn.py:522]: Readability suggestion, what do you think? Might be overly complicated.

```python
top_marker, deep_marker = 0xAAAA, 0xBBBB
deep_index = -(stack_index + 1)

stack = [0] * fork.max_stack_height()
stack[-1], stack[deep_index] = top_marker, deep_marker

code = Bytecode()
for value in stack:
    code += Op.PUSH2(value)

code += Op.SWAPN[stack_index]
stack[-1], stack[deep_index] = stack[deep_index], stack[-1]

storage = Storage()
code += Op.SSTORE(storage.store_next(stack.pop()), deep_marker)

while stack[-1] == 0:
    stack.pop()
    code += Op.POP

code += Op.SSTORE(storage.store_next(stack.pop()), top_marker)

contract_address = pre.deploy_contract(code=code)

tx = Transaction(to=contract_address, sender=pre.fund_eoa())

post = {contract_address: Account(storage=storage)}

state_test(pre=pre, post=post, tx=tx)
```

**Review by LouisTsai-Csie — CHANGES_REQUESTED** [review]
Some suggestion!

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py (hunk 6 above)
Comment (spencer-tb) [reply tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py]: Adopted with a small tweak in `2b3241cad0`.

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_exchange.py (hunk 7 above)
Comment (spencer-tb) [reply tests/amsterdam/eip8024_dupn_swapn_exchange/test_exchange.py:494]: Adopted with a small fix in `2482ee365c`.

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_vectors.py (hunk 5 above)
Comment (spencer-tb) [reply tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_vectors.py:819]: I added here as 9 other vectors in the file pin these the exact same way, and I think nice for safety

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_swapn.py (hunk 8 above)
Comment (spencer-tb) [reply tests/amsterdam/eip8024_dupn_swapn_exchange/test_swapn.py:522]: I didn't do this but I refactored in a different way: https://github.com/ethereum/execution-specs/pull/3224/commits/bc12faad61b3910ea4ad2e6e24a9fa41f08b6f7a

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py [hunk 9]
```diff
@@ -398,3 +408,113 @@ def test_dupn_with_dup1_sequence(
...
+    setup = Op.PUSH0 * Spec.MIN_STACK_INDEX
+    # The 0x5b landing pad sits 4 bytes past the setup: PUSH1,
+    # target, JUMP, then the DUPN opcode byte itself.
+    landing_pad = len(setup) + 4
+
+    code = setup
+    code += Op.PUSH1(landing_pad) + Op.JUMP
+    code += Op.DUPN[b"\x5b"]  # Jumped into, never executed.
+    code += Op.DUPN[Spec.MIN_STACK_INDEX]  # Executed after landing.
+    code += Op.SSTORE(0, 0x42)
```
Comment (marioevz) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py]:
```suggestion
    code += Op.SSTORE(
        0,
        Op.DUPN[Spec.MIN_STACK_INDEX],  # Executed after landing.
    )
```
Store the result of the execution, see complementary comment above.

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py [hunk 10]
```diff
@@ -398,3 +408,113 @@ def test_dupn_with_dup1_sequence(
...
+    because analysis is unchanged by EIP-8024, and execution runs a
+    second DUPN that duplicates the deepest of the 17 pushed items
+    before storing 0x42 over the canary. Wrong implementations diverge
+    observably: masking the immediate during analysis rejects the jump,
+    and mis-decoding the executed DUPN underflows; both abort and leave
+    the canary intact.
+    """
+    sender = pre.fund_eoa()
+
+    setup = Op.PUSH0 * Spec.MIN_STACK_INDEX
```
Comment (marioevz) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py]:
```suggestion
    setup = sum(Op.PUSH1[i] for i in range(Spec.MIN_STACK_INDEX, 0, -1))
```
To store something that is not zero.

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py [hunk 11]
```diff
@@ -398,3 +408,113 @@ def test_dupn_with_dup1_sequence(
...
+    code += Op.DUPN[b"\x5b"]  # Jumped into, never executed.
+    code += Op.DUPN[Spec.MIN_STACK_INDEX]  # Executed after landing.
+    code += Op.SSTORE(0, 0x42)
+    code += Op.STOP
+
+    contract_address = pre.deploy_contract(code=code, storage={0: 0xBA5E})
+
+    tx = Transaction(to=contract_address, sender=sender)
+
+    post = {contract_address: Account(storage={0: 0x42})}
```
Comment (marioevz) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py]:
```suggestion
    post = {contract_address: Account(storage={0: Spec.MIN_STACK_INDEX})}
```

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py [hunk 12]
```diff
@@ -398,3 +408,113 @@ def test_dupn_with_dup1_sequence(
...
+
+    Bytecode: PUSH0*17 PUSH1(21) JUMP DUPN[0x5b] DUPN[0x80] SSTORE
+
+    The jump lands on the first DUPN's 0x5b immediate, a valid JUMPDEST
+    because analysis is unchanged by EIP-8024, and execution runs a
+    second DUPN that duplicates the deepest of the 17 pushed items
+    before storing 0x42 over the canary. Wrong implementations diverge
+    observably: masking the immediate during analysis rejects the jump,
+    and mis-decoding the executed DUPN underflows; both abort and leave
+    the canary intact.
```
Comment (marioevz) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py]: The comment now has to be updated according to suggestions below. IMO these overly verbose comments are describing the code below almost line by line, we could tone it down a bit I think.

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_mainnet.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    )
+    contract = pre.deploy_contract(
+        code=code,
+        storage={0: 0xBA5E, 1: 0xBA5E, 2: 0xBA5E},
+    )
+    tx = Transaction(
+        ty=0x02,
+        to=contract,
+        sender=pre.fund_eoa(),
+        gas_limit=200_000,
```
Comment (marioevz) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_mainnet.py:71]: Thank you :D I don't think I've mentioned this before but it's nice to have prudent gas limits for mainnet tests, otherwise the sender requires a very high balance.

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_vectors.py [hunk 14]
```diff
@@ -786,6 +786,52 @@ def test_vector_exchange_invalid_0x52(
...
+    the 0x5b at an instruction boundary, accept the jump, and succeed —
+    diverging on bytecode that is valid before and after the fork.
+    """
+    sender = pre.fund_eoa()
+
+    # 00 PUSH1 0x05
+    # 02 JUMP
+    # 03 <opcode>
+    # 04 PUSH1 0x5b
+    code = Op.PUSH1(5) + Op.JUMP + opcode[b"\x60"] + Op.JUMPDEST
```
Comment (marioevz) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_eip_vectors.py]:
```suggestion
    code = Op.PUSH1(5) + Op.JUMP + opcode[bytes(Op.PUSH1)] + Op.JUMPDEST
```
Just makes the intent a tiny bit more obvious.

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_execution_contexts.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+            storage=EXPECTED_STORAGE,
+        ),
+    }
+
+    state_test(
+        env=Environment(),
+        pre=pre,
+        tx=tx,
+        post=post,
+    )
```
Comment (marioevz) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_execution_contexts.py]:
```suggestion
    state_test(pre=pre, tx=tx, post=post)
```

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_fork_transition.py
Comment (marioevz) [inline tests/amsterdam/eip8024_dupn_swapn_exchange/test_fork_transition.py:1]: We can add a new test to check that a `Op.JUMPDEST` in the data portion of any of the new opcodes works before and after the fork.

**Review by marioevz — COMMENTED** [review]
Looks great! It's nice to see the checklist to 100% 😄 

I left a couple of suggestions. Thanks!

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py (hunk 6 above)
Comment (LouisTsai-Csie) [reply tests/amsterdam/eip8024_dupn_swapn_exchange/test_dupn.py]: Thanks!

File: tests/amsterdam/eip8024_dupn_swapn_exchange/test_exchange.py (hunk 7 above)
Comment (LouisTsai-Csie) [reply tests/amsterdam/eip8024_dupn_swapn_exchange/test_exchange.py:494]: thanks

**Review by LouisTsai-Csie — APPROVED** [review]
I've confirmed all my suggestions are resolved! Thank you

**Comment by spencer-tb** [comment]
@marioevz these should all be addressed! Feel free to merge when you get the chance :)

**Review by marioevz — COMMENTED** [review]
Can we add a benchmark test for all the opcodes please?

Also, can we add a worst case jumpdest for a max-sized contract that is filled with jumpdest in the data section of the opcodes, and then we make a single jump to the end and end execution?

**Comment by spencer-tb** [comment]
> Also, can we add a worst case jumpdest for a max-sized contract that is filled with jumpdest in the data section of the opcodes, and then we make a single jump to the end and end execution?

Moved to: https://github.com/ethereum/execution-specs/pull/3267
